ogr/ogrsf_frmts/hana/ogrhanalayer.cpp

```diff
@@ -0,0 +1,955 @@
+/************************************************************************/
+
+OGRErr OGRHanaLayer::GetExtent(int iGeomField, OGREnvelope* extent, int force)
+{
+    if( iGeomField < 0 || iGeomField >= GetLayerDefn()->GetGeomFieldCount() ||
+        GetLayerDefn()->GetGeomFieldDefn(iGeomField)->GetType() == wkbNone )
+    {
+        extent->MinX = 0.0;
+        extent->MaxX = 0.0;
+        extent->MinY = 0.0;
+        extent->MaxY = 0.0;
+
+        if( iGeomField != 0 )
+        {
+            CPLError(CE_Failure, CPLE_AppDefined,
+                     "Invalid geometry field index : %d", iGeomField);
+        }
+        return OGRERR_FAILURE;
+    }
+
+    try
+    {
+        ReadGeometryExtent(iGeomField, extent);
+        return OGRERR_NONE;
+    }
+    catch (const std::exception& ex)
+    {
+        CPLString clmName =
+            (iGeomField < static_cast<int>(geomColumns_.size()))
+                ? geomColumns_[static_cast<std::size_t>(
+                                   geomColumns_.size())].name
+                : "unknown column";
+        CPLError(CE_Failure, CPLE_AppDefined,
+            "Unable to query extent of '%s' using fast method: %s",
+            clmName.c_str(), ex.what());
+    }
+
+    if( iGeomField == 0 )
+        return OGRLayer::GetExtent( extent, force );
+    else
+        return OGRLayer::GetExtent( iGeomField, extent, force);
+}
+
+/************************************************************************/
+/*                            GetFeatureCount()                          */
+/************************************************************************/
+
+GIntBig OGRHanaLayer::GetFeatureCount(CPL_UNUSED int force)
+{
+    GIntBig ret = FALSE;
+    CPLString sql = StringFormat(
+        "SELECT COUNT(*) FROM (%s) AS tmp", queryStatement_.c_str());
+    odbc::StatementRef stmt = dataSource_->CreateStatement();
+    odbc::ResultSetRef rs = stmt->executeQuery(sql.c_str());
+    if (rs->next())
+        ret = *rs->getLong(1);
+    rs->close();
+    return ret;
+}
+
+/************************************************************************/
+/*                         GetNextFeature()                             */
+/************************************************************************/
+
+OGRFeature* OGRHanaLayer::GetNextFeature()
+{
+    while (true)
+    {
+        OGRFeature* feature = GetNextFeatureInternal();
+        if (feature == nullptr)
+            return nullptr;
+
+        if ((m_poFilterGeom == nullptr
+             || FilterGeometry(feature->GetGeometryRef()))
+            && (m_poAttrQuery == nullptr || m_poAttrQuery->Evaluate(feature)))
+            return feature;
+
+        delete feature;
+    }
+}
+
+/************************************************************************/
+/*                           GetSpatialRef()                            */
+/************************************************************************/
+
+OGRSpatialReference* OGRHanaLayer::GetSpatialRef()
```

Comment:
implementing that method isn't needed if you SetSpatialRef() on created OGRGeometryFieldDefn.  And actually you should make sure that a user calling GetGeomFieldDefn(0)->GetSpatialRef() gets the same result as below. (didn't check if it is already the case. Lost in review...)

Reply:
Thanks for pointing this out. This function was removed in 0d769d4a7dcaab9327c6bcd69042ddb04decddba.